### PR TITLE
Start Redis in protected mode when started via ray.init().

### PR DESCRIPTION
**Pull request by robertnishihara**
Related to https://github.com/ray-project/ray/issues/2457.

This PR makes it so that when Ray is started via `ray.init()` (as opposed to via `ray start`) the Redis servers will be started in "protected mode" (which means that clients can only connect by connecting to `localhost`).

In practice, we actually connect redis clients by passing in the node IP address (not `localhost`), so I need to create a redis config file on the fly to allow both `localhost` and the node's actual IP address (it would have been nice to find a way to do this from the Python redis client, but I couldn't find one).

### CODE REVIEW

**Comment by robertnishihara** [comment]
@simon-mo, can you review this as well?

**Review by ericl — APPROVED** [review]

**Review by simon-mo — APPROVED** [review]
lgtm. This is great!

**Comment by ericl** [comment]
Looks like the test failure is the typical xray + pytorch crash.

**Comment by ericl** [comment]
Travis failures look unrelated.

File: python/ray/services.py [hunk 1]
```diff
@@ -553,13 +558,30 @@ def start_redis(node_ip_address,
...
 
 
+def _make_temp_redis_config(node_ip_address):
+    """Create a configuration file for Redis.
+
+    Args:
+        node_ip_address: The IP address of this node. This should not be
+            127.0.0.1.
+    """
+    redis_config_name = "/tmp/redis_conf{}".format(random_name())
```
Comment (mitar) [inline python/ray/services.py]: Nobody cleans up this file, no? :-)

This is really what [`NamedTemporaryFile`](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) was designed to do.

**Pull request merged**